File: runtimes/web/src/engine/background/BackgroundOriginConfig.ts
// BackgroundOriginConfig.ts — typed record for CSS `background-origin`.
// IR shape: array of { type: 'border-box'|'padding-box'|'content-box' }.

// One entry per layer.
export interface BackgroundOriginLayer {
  css: string;                                                        // CSS token for one layer
}

// Config holder.
export interface BackgroundOriginConfig {
  layers: BackgroundOriginLayer[];                                    // empty = unset
}

// IR property type.
export const BACKGROUND_ORIGIN_PROPERTY_TYPE = 'BackgroundOrigin' as const;
export type BackgroundOriginPropertyType = typeof BACKGROUND_ORIGIN_PROPERTY_TYPE;

File: runtimes/swiftui/Sources/StyleConverterRuntime/StyleEngine/background/BackgroundOriginConfig.swift
//
//  BackgroundOriginConfig.swift
//  StyleEngine/background — Phase 4.
//
//  CSS `background-origin` — which box the background positioning area
//  starts from (border-box / padding-box / content-box). IR shape is an
//  array of `{type: "<kind>"}` objects, one per layer. This property
//  only matters when `background-position` is non-zero; for a layer
//  that fills (gradient without position) the result is indistinguishable
//  from the default.
//

import Foundation

// Per-layer origin enum. Normalised from IR strings.
enum BackgroundOriginMode: Equatable {
    case borderBox
    case paddingBox
    case contentBox
}

struct BackgroundOriginConfig: Equatable {
    // All layers' origins, retained even though the applier is identity.
    var layers: [BackgroundOriginMode] = []
    // True when the property appeared in IR.
    var hasAny: Bool { !layers.isEmpty }
}
